### DISCUSSION TITLE
how is in_sample_residuals calculated?

### DISCUSSION THREAD

**Discussion by lucayapi** [JoaquinAmatRodrigo/skforecast#705] [General]
I would like to confirm how forecaster.in_sample_residuals is calculated.
$$\text{Residual}_i = y_i - \hat{y}_i$$
so the following code gives the adjusted values

```
lag = 1
forecaster = ForecasterAutoreg(
                         regressor=PoissonRegressor(),
                         lags=lag
                     )
forecaster.fit(y=train['y'])
predictions = forecaster.predict(steps=steps)
fitted_values = train['y'][lag:] - forecaster.in_sample_residuals
```

what's strange is that I can't find positive fitted values, so I use fish regression !
Does anyone have any solutions for me?

**Comment by JavierEscobarOrtiz** [comment]
Hello @lucayapi 

Yes, residuals are calculated as y_true - y_pred.

Maybe the `forecaster.in_sample_residuals` are quite high and the real values are close to 0, so the substraccion gets a value below 0?

To calculate in sample predictions you can also use the following code: (but it should give you the same result)

https://skforecast.org/latest/user_guides/autoregresive-forecaster#prediction-on-training-data

I recommend you check these tips that may help you:

[Avoid Negative Predictions in Forecasting](https://skforecast.org/latest/faq/non-negative-predictions)

[Intervals conditioned on predicted values (binned residuals)](https://skforecast.org/latest/user_guides/probabilistic-forecasting.html#intervals-conditioned-on-predicted-values-binned-residuals)

Let me know if any of these resources help you! 😄

**Comment by JoaquinAmatRodrigo** [comment]
Hi @lucayapi,
By fitted_values, do you mean the predicted values?
If so, the code you showed mixes in-sample residuals with out-sample predictions. Also, the in-sample residuals are a random sample with a maximum size of 200*n_bins. Therefore, they do not follow the same order as the training series.

**Reply by lucayapi** [reply]
I solved my problem by using 
https://skforecast.org/latest/user_guides/autoregresive-forecaster#prediction-on-training-data
By fitted_values, I was referring to the train prediction.

thank you !
